client.go

```diff
@@ -489,17 +491,40 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
 	var err error
 	for i := 0; i < maxHosts; i++ {
 		reader := bytes.NewReader(data)
-		// get a host from the cluster
-		host := c.cluster.Host()
-		if host == nil {
-			return nil, nil, ErrEmptyCluster
+		var host *URI
+
+		if useCoordinator {
+			c.coordinatorLock.RLock()
+			coordinatorURI := c.coordinatorURI
+			c.coordinatorLock.RUnlock()
+			if coordinatorURI == nil {
+				node, err := c.fetchCoordinatorNode()
+				if err != nil {
+					return nil, nil, errors.Wrap(err, "fetching coordinator node")
+				}
+				coordinatorURI = URIFromAddress(fmt.Sprintf("%s://%s:%d", node.Scheme, node.Host, node.Port))
+				c.coordinatorLock.Lock()
```

Comment:
I *think* this lock should be acquired before calling `fetchCoordinatorNode` to avoid multiple goroutines trying to fetch the coordinator simultaneously. This also needs to do a final check of c.coordinatorURI after acquriring the write lock in case it was set by another goroutine between releasing the read lock and acquiring the write lock.

Reply:
I've locked the mutex after fetching the coordinator node to prevent other goroutines block reading the coordinatorURI, but I am not sure my logic in that code is correct. I'm adding a test to see whether anything blocks.